source/postprocess/visualization/boundary_velocity_residual.cc

```diff
@@ -95,129 +66,12 @@ namespace aspect
-                                 "interpreted as the path in which the ASPECT source files were "
-                                 "located when ASPECT was compiled. This interpretation allows, "
-                                 "for example, to reference files located in the `data/' subdirectory "
-                                 "of ASPECT.");
-              prm.declare_entry ("Data file name", "current_day.gpml",
-                                 Patterns::Anything (),
-                                 "The file name of the input velocity as a GPlates model or an ascii data. "
-                                 "For the GPlates model, provide file in the same format as described "
-                                 "in the 'gplates' boundary velocity plugin. "
-                                 "For the ascii data, provide file in the same format as described in "
-                                 " 'ascii data' initial composition plugin." );
-              prm.declare_entry ("Scale factor", "1.",
-                                 Patterns::Double (),
-                                 "Scalar factor, which is applied to the model data. "
-                                 "You might want to use this to scale the input to a "
-                                 "reference model. Another way to use this factor is to "
-                                 "convert units of the input files. For instance, if you "
-                                 "provide velocities in cm/yr set this factor to 0.01.");
-              prm.declare_entry ("Use spherical unit vectors", "false",
-                                 Patterns::Bool (),
-                                 "Specify velocity as r, phi, and theta components "
-                                 "instead of x, y, and z. Positive velocities point up, east, "
-                                 "and north (in 3D) or out and clockwise (in 2D). "
-                                 "This setting only makes sense for spherical geometries."
-                                 "GPlates data is always interpreted to be in east and north directions "
-                                 "and is not affected by this parameter.");
-              prm.declare_entry ("Use ascii data", "false",
-                                 Patterns::Bool (),
-                                 "Use ascii data files (e.g., GPS) for computing residual velocities "
-                                 "instead of GPlates data.");
-            }
-            prm.leave_subsection();
-          }
-          prm.leave_subsection();
-        }
-        prm.leave_subsection();
+          for (unsigned int q=0; q<computed_quantities.size(); ++q)
+            for (unsigned int d = 0; d < dim; ++d)
+              computed_quantities[q](d) = boundary_velocity_residual_statistics.evaluate(input_data.evaluation_points[q])[d] -
+                                          input_data.solution_values[q][d] * velocity_scaling_factor;
```

Comment:
This is correct, but consider what is happening here: You call the `evaluate` function oncer per dimension, but then only use the `d`'th component and ignore the rest, then repeat this computation two more times. It likely does not matter here, but for other cases you want to optimize this in the following way. You can do this because the call to `evaluate` does not depend on `d`.

```suggestion
{
            const Tensor<1,dim> data_velocity = boundary_velocity_residual_statistics.evaluate(input_data.evaluation_points[q]);
            for (unsigned int d = 0; d < dim; ++d)
              computed_quantities[q](d) = data_velocity[d] -
                                          input_data.solution_values[q][d] * velocity_scaling_factor;
 }
```